compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt

```diff
@@ -82,18 +84,148 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
+                        irNotEquals(untilArgExpression, minValueIrConst)
+                    it == minValueAsLong ->
+                        // Hardcode "false" as additional condition so that the progression is considered empty.
+                        // The entire lowered loop becomes a candidate for dead code elimination, depending on backend.
+                        irFalse()
+                    else ->
+                        // We know that untilArg != MIN_VALUE, so the additional condition is not necessary.
+                        null
+                }
+            }
+
             ProgressionHeaderInfo(
                 data,
-                first = expression.extensionReceiver!!,
-                last = expression.getValueArgument(0)!!,
+                first = first,
+                last = last,
                 step = irInt(1),
-                isLastInclusive = false
+                canOverflow = false,
+                additionalVariables = additionalVariables,
+                additionalNotEmptyCondition = additionalNotEmptyCondition,
+                direction = ProgressionDirection.INCREASING
             )
         }
+
+    private fun isAdditionalNotEmptyConditionNeeded(receiverType: IrType, argType: IrType): Boolean {
+        val receiverKotlinType = receiverType.toKotlinType()
+        val argKotlinType = argType.toKotlinType()
+
+        // Here are the available `until` extension functions:
+        //
+        //   infix fun Char.until(to: Char): CharRange
+        //   infix fun Byte.until(to: Byte): IntRange
+        //   infix fun Byte.until(to: Short): IntRange
+        //   infix fun Byte.until(to: Int): IntRange
+        //   infix fun Byte.until(to: Long): LongRange
+        //   infix fun Short.until(to: Byte): IntRange
+        //   infix fun Short.until(to: Short): IntRange
+        //   infix fun Short.until(to: Int): IntRange
+        //   infix fun Short.until(to: Long): LongRange
+        //   infix fun Int.until(to: Byte): IntRange
+        //   infix fun Int.until(to: Short): IntRange
+        //   infix fun Int.until(to: Int): IntRange
+        //   infix fun Int.until(to: Long): LongRange
+        //   infix fun Long.until(to: Byte): LongRange
+        //   infix fun Long.until(to: Short): LongRange
+        //   infix fun Long.until(to: Int): LongRange
+        //   infix fun Long.until(to: Long): LongRange
+        //
+        // The combinations where the range element type is strictly larger than the argument type do NOT need the additional condition.
+        // In such combinations, there is no possibility of underflow when the argument (casted to the range element type) is decremented.
+        // For unexpected combinations that currently don't exist (e.g., Int until Char), we assume the check is needed to be safe.
+        // TODO: Include unsigned types
+        return with(context.irBuiltIns) {
+            when (receiverKotlinType) {
+                char -> true
```

Comment:
Why not use the IrTypes directly and compare against charType, byteType, etc.?

Reply:
My recollection could be wrong, but when I started working on for-loops, I was not able to compare IrTypes directly. Anyway, updated to use IrTypes, and also updated elsewhere in the code where I was using KotlinType.